Index: modules/core/org.jowidgets.api/src/main/java/org/jowidgets/api/widgets/IContainer.java
===================================================================
--- modules/core/org.jowidgets.api/src/main/java/org/jowidgets/api/widgets/IContainer.java	(revision 4173)
+++ modules/core/org.jowidgets.api/src/main/java/org/jowidgets/api/widgets/IContainer.java	(working copy)
@@ -118,6 +118,17 @@
 		Object layoutConstraints);
 
 	/**
+	 * Creates and adds an control to this container.
+	 * 
+	 * @param <WIDGET_TYPE> The type of the widget that is created, added and returned
+	 * @param index The index at which the control should be added in the container
+	 * @param descriptor The descriptor that describes the control to add
+	 * 
+	 * @return the created and added control
+	 */
+	<WIDGET_TYPE extends IControl> WIDGET_TYPE add(int index, IWidgetDescriptor<? extends WIDGET_TYPE> descriptor);
+
+	/**
 	 * Creates and adds an control to this container
 	 * 
 	 * @param <WIDGET_TYPE> The type of the widget that is created, added and returned
@@ -130,6 +141,17 @@
 	<WIDGET_TYPE extends IControl> WIDGET_TYPE add(int index, ICustomWidgetCreator<WIDGET_TYPE> creator, Object layoutConstraints);
 
 	/**
+	 * Creates and adds an control to this container
+	 * 
+	 * @param <WIDGET_TYPE> The type of the widget that is created, added and returned
+	 * @param index The index at which the control should be added in the container
+	 * @param creator The creator that creates the control
+	 * 
+	 * @return the created and added control
+	 */
+	<WIDGET_TYPE extends IControl> WIDGET_TYPE add(int index, ICustomWidgetCreator<WIDGET_TYPE> creator);
+
+	/**
 	 * Creates and adds an control to this container.
 	 * 
 	 * @param <WIDGET_TYPE> The type of the widget that is created, added and returned
Index: modules/core/org.jowidgets.api/src/main/java/org/jowidgets/api/widgets/IControl.java
===================================================================
--- modules/core/org.jowidgets.api/src/main/java/org/jowidgets/api/widgets/IControl.java	(revision 4173)
+++ modules/core/org.jowidgets.api/src/main/java/org/jowidgets/api/widgets/IControl.java	(working copy)
@@ -36,6 +36,14 @@
 
 public interface IControl extends IComponent, IControlCommon, IParentObservable<IContainer> {
 
+	/**
+	 * Sets the parent of the container.
+	 * 
+	 * If the parent is already set and the control is not reparentable an
+	 * IllegalStateException will be thrown.
+	 * 
+	 * @param parent The parent to set
+	 */
 	void setParent(IContainer parent);
 
 	@Override
@@ -44,8 +52,18 @@
 	@Override
 	IControl getRoot();
 
+	/**
+	 * Gets the drag source of the control
+	 * 
+	 * @return The drag source, never null
+	 */
 	IDragSource getDragSource();
 
+	/**
+	 * Gets the drop target of the control
+	 * 
+	 * @return The drop target, never null
+	 */
 	IDropTarget getDropTarget();
 
 	void setMinSize(final Dimension minSize);
Index: modules/core/org.jowidgets.common/src/main/java/org/jowidgets/common/widgets/factory/ICustomWidgetCreator.java
===================================================================
--- modules/core/org.jowidgets.common/src/main/java/org/jowidgets/common/widgets/factory/ICustomWidgetCreator.java	(revision 4173)
+++ modules/core/org.jowidgets.common/src/main/java/org/jowidgets/common/widgets/factory/ICustomWidgetCreator.java	(working copy)
@@ -31,6 +31,13 @@
 
 public interface ICustomWidgetCreator<WIDGET_TYPE extends IWidgetCommon> {
 
+	/**
+	 * Creates a widget
+	 * 
+	 * @param widgetFactory The widget factory that can be used
+	 * 
+	 * @return The created widget, never null
+	 */
 	WIDGET_TYPE create(ICustomWidgetFactory widgetFactory);
 
 }
Index: modules/core/org.jowidgets.impl/src/main/java/org/jowidgets/impl/base/delegate/ContainerDelegate.java
===================================================================
--- modules/core/org.jowidgets.impl/src/main/java/org/jowidgets/impl/base/delegate/ContainerDelegate.java	(revision 4177)
+++ modules/core/org.jowidgets.impl/src/main/java/org/jowidgets/impl/base/delegate/ContainerDelegate.java	(working copy)
@@ -298,11 +298,21 @@
 
 	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(
 		final int index,
+		final IWidgetDescriptor<? extends WIDGET_TYPE> descriptor) {
+		return add(index, descriptor, null);
+	}
+
+	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(
+		final int index,
 		final ICustomWidgetCreator<WIDGET_TYPE> creator,
 		final Object layoutConstraints) {
 		return add(Integer.valueOf(index), creator, layoutConstraints);
 	}
 
+	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(final int index, final ICustomWidgetCreator<WIDGET_TYPE> creator) {
+		return add(Integer.valueOf(index), creator, null);
+	}
+
 	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(
 		final IWidgetDescriptor<? extends WIDGET_TYPE> descriptor,
 		final Object layoutConstraints) {
Index: modules/core/org.jowidgets.impl/src/main/java/org/jowidgets/impl/widgets/basic/ContainerImpl.java
===================================================================
--- modules/core/org.jowidgets.impl/src/main/java/org/jowidgets/impl/widgets/basic/ContainerImpl.java	(revision 4173)
+++ modules/core/org.jowidgets.impl/src/main/java/org/jowidgets/impl/widgets/basic/ContainerImpl.java	(working copy)
@@ -202,6 +202,13 @@
 	@Override
 	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(
 		final int index,
+		final IWidgetDescriptor<? extends WIDGET_TYPE> descriptor) {
+		return containerDelegate.add(index, descriptor);
+	}
+
+	@Override
+	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(
+		final int index,
 		final ICustomWidgetCreator<WIDGET_TYPE> creator,
 		final Object layoutConstraints) {
 		return containerDelegate.add(index, creator, layoutConstraints);
@@ -208,6 +215,11 @@
 	}
 
 	@Override
+	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(final int index, final ICustomWidgetCreator<WIDGET_TYPE> creator) {
+		return containerDelegate.add(index, creator);
+	}
+
+	@Override
 	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(
 		final IWidgetDescriptor<? extends WIDGET_TYPE> descriptor,
 		final Object layoutConstraints) {
Index: modules/core/org.jowidgets.impl/src/main/java/org/jowidgets/impl/widgets/basic/FrameImpl.java
===================================================================
--- modules/core/org.jowidgets.impl/src/main/java/org/jowidgets/impl/widgets/basic/FrameImpl.java	(revision 4173)
+++ modules/core/org.jowidgets.impl/src/main/java/org/jowidgets/impl/widgets/basic/FrameImpl.java	(working copy)
@@ -226,6 +226,13 @@
 	@Override
 	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(
 		final int index,
+		final IWidgetDescriptor<? extends WIDGET_TYPE> descriptor) {
+		return containerDelegate.add(index, descriptor);
+	}
+
+	@Override
+	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(
+		final int index,
 		final ICustomWidgetCreator<WIDGET_TYPE> creator,
 		final Object layoutConstraints) {
 		return containerDelegate.add(index, creator, layoutConstraints);
@@ -232,6 +239,11 @@
 	}
 
 	@Override
+	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(final int index, final ICustomWidgetCreator<WIDGET_TYPE> creator) {
+		return containerDelegate.add(index, creator);
+	}
+
+	@Override
 	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(
 		final IWidgetDescriptor<? extends WIDGET_TYPE> descriptor,
 		final Object layoutConstraints) {
Index: modules/core/org.jowidgets.impl/src/main/java/org/jowidgets/impl/widgets/basic/PopupDialogImpl.java
===================================================================
--- modules/core/org.jowidgets.impl/src/main/java/org/jowidgets/impl/widgets/basic/PopupDialogImpl.java	(revision 4173)
+++ modules/core/org.jowidgets.impl/src/main/java/org/jowidgets/impl/widgets/basic/PopupDialogImpl.java	(working copy)
@@ -266,6 +266,13 @@
 	@Override
 	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(
 		final int index,
+		final IWidgetDescriptor<? extends WIDGET_TYPE> descriptor) {
+		return containerDelegate.add(index, descriptor);
+	}
+
+	@Override
+	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(
+		final int index,
 		final ICustomWidgetCreator<WIDGET_TYPE> creator,
 		final Object layoutConstraints) {
 		return containerDelegate.add(index, creator, layoutConstraints);
@@ -272,6 +279,11 @@
 	}
 
 	@Override
+	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(final int index, final ICustomWidgetCreator<WIDGET_TYPE> creator) {
+		return containerDelegate.add(index, creator);
+	}
+
+	@Override
 	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(
 		final IWidgetDescriptor<? extends WIDGET_TYPE> descriptor,
 		final Object layoutConstraints) {
Index: modules/core/org.jowidgets.impl/src/main/java/org/jowidgets/impl/widgets/basic/ScrollCompositeImpl.java
===================================================================
--- modules/core/org.jowidgets.impl/src/main/java/org/jowidgets/impl/widgets/basic/ScrollCompositeImpl.java	(revision 4173)
+++ modules/core/org.jowidgets.impl/src/main/java/org/jowidgets/impl/widgets/basic/ScrollCompositeImpl.java	(working copy)
@@ -181,6 +181,13 @@
 	@Override
 	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(
 		final int index,
+		final IWidgetDescriptor<? extends WIDGET_TYPE> descriptor) {
+		return containerDelegate.add(index, descriptor);
+	}
+
+	@Override
+	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(
+		final int index,
 		final ICustomWidgetCreator<WIDGET_TYPE> creator,
 		final Object layoutConstraints) {
 		return containerDelegate.add(index, creator, layoutConstraints);
@@ -187,6 +194,11 @@
 	}
 
 	@Override
+	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(final int index, final ICustomWidgetCreator<WIDGET_TYPE> creator) {
+		return containerDelegate.add(index, creator);
+	}
+
+	@Override
 	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(
 		final IWidgetDescriptor<? extends WIDGET_TYPE> descriptor,
 		final Object layoutConstraints) {
Index: modules/core/org.jowidgets.impl/src/main/java/org/jowidgets/impl/widgets/basic/TabItemImpl.java
===================================================================
--- modules/core/org.jowidgets.impl/src/main/java/org/jowidgets/impl/widgets/basic/TabItemImpl.java	(revision 4173)
+++ modules/core/org.jowidgets.impl/src/main/java/org/jowidgets/impl/widgets/basic/TabItemImpl.java	(working copy)
@@ -467,6 +467,14 @@
 	@Override
 	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(
 		final int index,
+		final IWidgetDescriptor<? extends WIDGET_TYPE> descriptor) {
+		checkDetached();
+		return containerDelegate.add(index, descriptor);
+	}
+
+	@Override
+	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(
+		final int index,
 		final ICustomWidgetCreator<WIDGET_TYPE> creator,
 		final Object layoutConstraints) {
 		checkDetached();
@@ -490,6 +498,12 @@
 	}
 
 	@Override
+	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(final int index, final ICustomWidgetCreator<WIDGET_TYPE> creator) {
+		checkDetached();
+		return containerDelegate.add(index, creator);
+	}
+
+	@Override
 	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(final IWidgetDescriptor<? extends WIDGET_TYPE> descriptor) {
 		checkDetached();
 		return containerDelegate.add(descriptor);
Index: modules/core/org.jowidgets.impl/src/main/java/org/jowidgets/impl/widgets/composed/ExpandCompositeImpl.java
===================================================================
--- modules/core/org.jowidgets.impl/src/main/java/org/jowidgets/impl/widgets/composed/ExpandCompositeImpl.java	(revision 4173)
+++ modules/core/org.jowidgets.impl/src/main/java/org/jowidgets/impl/widgets/composed/ExpandCompositeImpl.java	(working copy)
@@ -348,7 +348,19 @@
 	}
 
 	@Override
+	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(final int index, final ICustomWidgetCreator<WIDGET_TYPE> creator) {
+		return content.add(index, creator);
+	}
+
+	@Override
 	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(
+		final int index,
+		final IWidgetDescriptor<? extends WIDGET_TYPE> descriptor) {
+		return content.add(index, descriptor);
+	}
+
+	@Override
+	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(
 		final IWidgetDescriptor<? extends WIDGET_TYPE> descriptor,
 		final Object layoutConstraints) {
 		return content.add(descriptor, layoutConstraints);
Index: modules/core/org.jowidgets.impl/src/main/java/org/jowidgets/impl/widgets/composed/InputCompositeImpl.java
===================================================================
--- modules/core/org.jowidgets.impl/src/main/java/org/jowidgets/impl/widgets/composed/InputCompositeImpl.java	(revision 4173)
+++ modules/core/org.jowidgets.impl/src/main/java/org/jowidgets/impl/widgets/composed/InputCompositeImpl.java	(working copy)
@@ -434,6 +434,13 @@
 
 	@Override
 	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(
+		final int index,
+		final IWidgetDescriptor<? extends WIDGET_TYPE> descriptor) {
+		return add(index, descriptor, null);
+	}
+
+	@Override
+	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(
 		final IWidgetDescriptor<? extends WIDGET_TYPE> descriptor,
 		final Object layoutConstraints) {
 
@@ -453,6 +460,11 @@
 	}
 
 	@Override
+	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(final int index, final ICustomWidgetCreator<WIDGET_TYPE> creator) {
+		return add(index, creator, null);
+	}
+
+	@Override
 	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(final IWidgetDescriptor<? extends WIDGET_TYPE> descriptor) {
 		final WIDGET_TYPE result = innerComposite.add(descriptor);
 		afterAdd(null, result);
Index: modules/core/org.jowidgets.impl/src/main/java/org/jowidgets/impl/widgets/composed/LabelImpl.java
===================================================================
--- modules/core/org.jowidgets.impl/src/main/java/org/jowidgets/impl/widgets/composed/LabelImpl.java	(revision 4173)
+++ modules/core/org.jowidgets.impl/src/main/java/org/jowidgets/impl/widgets/composed/LabelImpl.java	(working copy)
@@ -137,11 +137,15 @@
 
 	@Override
 	public void addPopupDetectionListener(final IPopupDetectionListener listener) {
+		textLabelWidget.addPopupDetectionListener(listener);
+		iconWidget.addPopupDetectionListener(listener);
 		composite.addPopupDetectionListener(listener);
 	}
 
 	@Override
 	public void removePopupDetectionListener(final IPopupDetectionListener listener) {
+		textLabelWidget.removePopupDetectionListener(listener);
+		iconWidget.removePopupDetectionListener(listener);
 		composite.removePopupDetectionListener(listener);
 	}
 
Index: modules/core/org.jowidgets.tools/src/main/java/org/jowidgets/tools/powo/Container.java
===================================================================
--- modules/core/org.jowidgets.tools/src/main/java/org/jowidgets/tools/powo/Container.java	(revision 4173)
+++ modules/core/org.jowidgets.tools/src/main/java/org/jowidgets/tools/powo/Container.java	(working copy)
@@ -357,6 +357,13 @@
 		}
 	}
 
+	@Override
+	public <M_WIDGET_TYPE extends IControl> M_WIDGET_TYPE add(
+		final int index,
+		final IWidgetDescriptor<? extends M_WIDGET_TYPE> descriptor) {
+		return add(index, descriptor, null);
+	}
+
 	@SuppressWarnings({"unchecked", "rawtypes"})
 	@Override
 	public final <M_WIDGET_TYPE extends IControl> M_WIDGET_TYPE add(
@@ -508,6 +515,11 @@
 	}
 
 	@Override
+	public <M_WIDGET_TYPE extends IControl> M_WIDGET_TYPE add(final int index, final ICustomWidgetCreator<M_WIDGET_TYPE> creator) {
+		return add(index, creator, null);
+	}
+
+	@Override
 	public final <M_WIDGET_TYPE extends IControl> M_WIDGET_TYPE add(
 		final ICustomWidgetCreator<M_WIDGET_TYPE> creator,
 		final Object layoutConstraints) {
Index: modules/core/org.jowidgets.tools/src/main/java/org/jowidgets/tools/widgets/wrapper/ContainerWrapper.java
===================================================================
--- modules/core/org.jowidgets.tools/src/main/java/org/jowidgets/tools/widgets/wrapper/ContainerWrapper.java	(revision 4173)
+++ modules/core/org.jowidgets.tools/src/main/java/org/jowidgets/tools/widgets/wrapper/ContainerWrapper.java	(working copy)
@@ -126,6 +126,13 @@
 	@Override
 	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(
 		final int index,
+		final IWidgetDescriptor<? extends WIDGET_TYPE> descriptor) {
+		return getWidget().add(index, descriptor);
+	}
+
+	@Override
+	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(
+		final int index,
 		final ICustomWidgetCreator<WIDGET_TYPE> creator,
 		final Object layoutConstraints) {
 		return getWidget().add(index, creator, layoutConstraints);
@@ -132,6 +139,11 @@
 	}
 
 	@Override
+	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(final int index, final ICustomWidgetCreator<WIDGET_TYPE> creator) {
+		return getWidget().add(index, creator);
+	}
+
+	@Override
 	public <WIDGET_TYPE extends IControl> WIDGET_TYPE add(
 		final IWidgetDescriptor<? extends WIDGET_TYPE> descriptor,
 		final Object layoutConstraints) {
